library/core/src/main/java/com/google/android/exoplayer2/audio/AudioCapabilitiesReceiver.java

```diff
@@ -91,14 +108,17 @@ public AudioCapabilities register() {
     if (externalSurroundSoundSettingObserver != null) {
       externalSurroundSoundSettingObserver.register();
     }
-    Intent stickyIntent = null;
     if (receiver != null) {
       IntentFilter intentFilter = new IntentFilter(AudioManager.ACTION_HDMI_AUDIO_PLUG);
       stickyIntent =
           context.registerReceiver(
               receiver, intentFilter, /* broadcastPermission= */ null, handler);
     }
     audioCapabilities = AudioCapabilities.getCapabilities(context, stickyIntent);
+
+    mediaRouter.addCallback(mSelector, mediaRouterCallback,
+            MediaRouter.CALLBACK_FLAG_REQUEST_DISCOVERY);
```

Comment:
Would using `CALLBACK_FLAG_REQUEST_DISCOVERY` not miss some state change, leading to erroneous capabilities?

> The CALLBACK_FLAG_REQUEST_DISCOVERY flag should be used while the application is in the foreground to request that passive discovery be performed if there are sufficient resources to allow continuous passive discovery. On low-RAM devices this flag will be ignored to conserve resources.
https://developer.android.com/reference/androidx/mediarouter/media/MediaRouter#addCallback(androidx.mediarouter.media.MediaRouteSelector,%20androidx.mediarouter.media.MediaRouter.Callback,%20int)

Reply:
Thanks @krocard for the feedback. We decided to use CALLBACK_FLAG_REQUEST_DISCOVERY based on https://developer.android.com/reference/androidx/mediarouter/media/MediaRouter#addCallback(androidx.mediarouter.media.MediaRouteSelector,%20androidx.mediarouter.media.MediaRouter.Callback,%20int)

CALLBACK_FLAG_FORCE_DISCOVERY seem to have significant performance impact.

In our experiment we are being notified when BlueTooth speaker connects without having to scan. i.e. even if the BT speaker connected earlier, when App comes up, the notification arrives here and we are able to set the PCM_ONLY Audioapabilities.

Reply:
If the capabilities are eventually consistent with `CALLBACK_FLAG_REQUEST_DISCOVERY`, it should indeed be enough. I do wonder though when ExoPlayer is playing in a foreground service how long the callback can be delayed.
If it is by more than a few seconds users are going to consider that a bug.